Marlin/src/HAL/STM32/platdefs/MARLIN_ARTILLERY_RUBY.h

```diff
@@ -0,0 +1,27 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2023 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+
+#ifndef NUM_ANALOG_FIRST
+  #error "NUM_ANALOG_FIRST is not defined??"
```

Comment:
So far none of the CI tests hit this error. So perhaps `variant.h` is taking care of this for us already.

Reply:
That is not true. In my testing the build target STM32F103RC_btt failed when PINS_DEBUGGING was enabled, which is part of the official Marlin FW CI tests (https://github.com/quiret/Marlin/blob/bugfix-2.1.x-tcupdate/buildroot/tests/STM32F103RC_btt).

https://github.com/quiret/Marlin/blob/bugfix-2.1.x-tcupdate/buildroot/share/PlatformIO/variants/MARLIN_F103Rx/variant.h#L90

Reply:
> It failed specifically because variant.h did not have `NUM_ANALOG_FIRST` defined.

In this case we should add `NUM_ANALOG_FIRST` to `variant.h`. I would very much like to avoid adding a set of `platdefs` headers that will just be another maintenance headache. I want to move in the direction of "more elegant" and "more organized" and this new set of headers adds to the scattered nature of our info.

In the long run towards more code modernization, I would like to be able to make a pins layer that allows pins to be variable at runtime –as an option– which will mean wrapping pins up in some kind of template class and using one of the flavors of "FastIO" that can allow for this.

It's one of the many dozens of things on the "make Marlin more modern" TODO list. The more I can get through the pile of existing PRs and fix the existing bugs, the more time this should leave open to start on a new experimental branch to redesign various aspects of the firmware.

Reply:
@thinkyhead You make a fair point! I have listened to you interview with Kersey Fabrications where you mentioned your vision for Marlin. It is correct that more restrictive statements in the form of preprocessor definitions contradict your vision of a accessible Marlin firmware, even to the point of re-assigning pins at runtime. You have peaked my interest and I would like to work on code that matches your idea!

Reply:
So far I don't know which variants are missing `NUM_ANALOG_FIRST` and breaking or hitting that `#error` when we try to compile them.

I'm glad to hear your interest is piqued! I actually have a small pile of planned fixes for the pins files, and the first one is to move the LCD / controller pin defines to their own files, so that every board can potentially support every LCD / controller. And after that I want to look at wrapping the pins definitions in a template or class or namespace, depending on what works best. I kind of like the neatness of namespaces, which works very well for our language definitions. The one thing that makes #defines useful now is that we can hide code from the compiler when a pin is not defined, whereas a template, class, or namespace will require code that isn't going to be used to be compile-able, and that is not always possible. As with anything in Marlin, things we can do gradually tend to get done, whereas the huge refactors require too much concentrated effort and can't get done without neglecting day-to-day patches.